File: apps/api/app/models/schedule.py
from datetime import datetime, time
from typing import TYPE_CHECKING

from sqlalchemy import CheckConstraint, DateTime, ForeignKey, Integer, String, Time, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base

if TYPE_CHECKING:
    from app.models.course import Course


class ScheduleEntry(Base):
    __tablename__ = "schedule_entries"
    __table_args__ = (
        CheckConstraint(
            "weekday >= 0 AND weekday <= 6",
            name="ck_schedule_entries_weekday",
        ),
        CheckConstraint(
            "start_time < end_time",
            name="ck_schedule_entries_time_range",
        ),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    course_id: Mapped[int] = mapped_column(
        ForeignKey("courses.id", ondelete="CASCADE"),
        index=True,
    )
    weekday: Mapped[int] = mapped_column(Integer, index=True)
    start_time: Mapped[time] = mapped_column(Time)
    end_time: Mapped[time] = mapped_column(Time)
    room: Mapped[str | None] = mapped_column(String(50), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    course: Mapped["Course"] = relationship(back_populates="schedule_entries")

File: apps/web/src/types/schedule.ts
export type ScheduleCourse = {
  id: number
  code: string
  name: string
  color: string
}

export type ScheduleEntry = {
  id: number
  course_id: number
  weekday: number
  start_time: string
  end_time: string
  room: string | null
  created_at: string
  course: ScheduleCourse
}

export type ScheduleEntryInput = {
  course_id: number
  weekday: number
  start_time: string
  end_time: string
  room: string | null
}

export type ScheduleEntryUpdateInput = Partial<ScheduleEntryInput>
